crates/net/network/src/transactions/validation.rs

```diff
@@ -0,0 +1,367 @@
+            debug!(target: "net::eth-wire",
+                ty=ty,
+                size=size,
+                hash=%hash,
+                max_encoded_tx_length=tx_type.max_encoded_tx_length(),
+                network=%Self,
+                "invalid tx size in eth68 announcement"
+            );
+
+            return ValidationOutcome::Ignore
+        }
+
+        ValidationOutcome::Fetch
+    }
+}
+
+impl FilterAnnouncement68 for LayerOne
+where
+    Self: ValidateTx68,
+{
+    fn filter_valid_entries(&self, msg: &mut NewPooledTransactionHashes68) -> FilterOutcome {
+        trace!(target: "net::tx::validation",
+            types=?msg.types,
+            sizes=?msg.sizes,
+            hashes=?msg.hashes,
+            network=%Self,
+            "validating eth68 announcement data.."
+        );
+
+        // 1. checks if the announcement is empty
+        if msg.hashes.is_empty() {
+            debug!(target: "net::tx",
+                network=%Self,
+                "empty announcement"
+            );
+            return FilterOutcome::ReportPeer
+        }
+
+        let mut should_report_peer = false;
+        let mut indices_to_remove = vec![];
+
+        for (i, (&ty, &hash, &size)) in izip!(&msg.types, &msg.hashes, &msg.sizes).enumerate() {
+            match self.should_fetch(ty, hash, size) {
+                ValidationOutcome::Fetch => (),
+                ValidationOutcome::Ignore => indices_to_remove.push(i),
+                ValidationOutcome::ReportPeer => {
+                    indices_to_remove.push(i);
+                    should_report_peer = true;
+                }
+            }
+        }
+
+        for (i, &index) in indices_to_remove.iter().rev().enumerate() {
+            let index = index.saturating_sub(i);
+
+            msg.hashes.remove(index);
+            msg.types.remove(index);
+            msg.sizes.remove(index);
+        }
+
+        let mut hashes_sorted = msg.hashes.clone();
+        hashes_sorted.sort();
+
+        for hash_pair in hashes_sorted.windows(2) {
+            if hash_pair[0] == hash_pair[1] {
+                let index = msg.hashes.iter().position(|&hash| hash == hash_pair[0]);
+                if let Some(i) = index {
+                    debug!(target: "net::eth-wire",
+                        ty=msg.types[i],
+                        size=msg.sizes[i],
+                        hash=%msg.hashes[i],
+                        network=%Self,
+                        "duplicate tx hash in eth68 announcement"
+                    );
+
+                    msg.hashes.remove(i);
+                    msg.types.remove(i);
+                    msg.sizes.remove(i);
+                }
+
+                should_report_peer = true;
+            }
+        }
+
+        if should_report_peer {
+            return FilterOutcome::ReportPeer
+        }
+
+        FilterOutcome::Ok
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::str::FromStr;
+
+    use reth_primitives::B256;
+
```

Comment:
```suggestion
```

Reply:
fyi, this is the layout with which rust-analyzer imports missing deps when you let it handle it
```rust
use std;

use other_crates;

use crate;

use super;
```